server/src/main/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesAction.java

```diff
@@ -0,0 +1,115 @@
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+public class TransportGetDiscoveredNodesAction extends TransportAction<GetDiscoveredNodesRequest, GetDiscoveredNodesResponse> {
+
+    @Nullable // TODO make this not nullable
+    private final Coordinator coordinator;
+    private final TransportService transportService;
+
+    @Inject
+    public TransportGetDiscoveredNodesAction(Settings settings, ActionFilters actionFilters, TransportService transportService,
+                                             Discovery discovery) {
+        super(settings, GetDiscoveredNodesAction.NAME, actionFilters, transportService.getTaskManager());
+        this.transportService = transportService;
+        if (discovery instanceof Coordinator) {
+            coordinator = (Coordinator) discovery;
+        } else {
+            coordinator = null;
+        }
+    }
+
+    @Override
+    protected void doExecute(Task task, GetDiscoveredNodesRequest request, ActionListener<GetDiscoveredNodesResponse> listener) {
+        if (coordinator == null) { // TODO remove when not nullable
+            throw new IllegalStateException("cannot execute a Zen2 action if not using Zen2");
+        }
+
+        final DiscoveryNode localNode = transportService.getLocalNode();
+        assert localNode != null;
+        if (localNode.isMasterNode() == false) {
+            throw new ElasticsearchException("this node is not master-eligible");
+        }
+
+        final AtomicBoolean responseSent = new AtomicBoolean();
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+        final long timeoutMillis = request.getTimeout().millis();
+        assert timeoutMillis >= 0 : timeoutMillis;
+
+        final Consumer<Iterable<DiscoveryNode>> respondIfRequestSatisfied = new Consumer<Iterable<DiscoveryNode>>() {
+            @Override
+            public void accept(Iterable<DiscoveryNode> nodes) {
+                final Set<DiscoveryNode> nodesSet = new LinkedHashSet<>();
+                nodesSet.add(localNode);
+                nodes.forEach(nodesSet::add);
+                if (nodesSet.size() >= request.getMinimumNodeCount() && responseSent.compareAndSet(false, true)) {
+                    listener.onResponse(new GetDiscoveredNodesResponse(nodesSet));
+                    countDownLatch.countDown();
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "waiting for " + request;
+            }
+        };
+
+        try (Releasable ignored = coordinator.withDiscoveryListener(respondIfRequestSatisfied)) {
+            respondIfRequestSatisfied.accept(coordinator.getFoundPeers());
+            if (timeoutMillis > 0 && countDownLatch.await(timeoutMillis, TimeUnit.MILLISECONDS)) {
```

Comment:
which thread is this executed on? Can we make this async instead of blocking here?

Reply:
As far as I can tell, it's executed on the thread that the client invokes it from. It's also only ever executed locally.

Reply:
Made async in d4818d621d081e20bdce329fa099f2b2641d84f0